### PR TITLE
Cilium CA: Autogenerate Cilium CA if it is missing

### PR DESCRIPTION
**Pull request by soggiest**
Per the [Setting Up Hubble Observability](https://docs.cilium.io/en/stable/gettingstarted/hubble_setup/#hubble-setup) and the [Setting Up Clustermesh](https://docs.cilium.io/en/stable/gettingstarted/clustermesh/clustermesh/#clustermesh) documentation users should expect the `cilium hubble enable` and `cilium clustermesh enable` commands to work as described without having to pass the `--create-ca` flag. 

This PR:
* Splits the `client.GetSecrets` function out of the `LoadCAFromK8s` function in `internal/certs/certs.go` into a new `GetCASecret` function.
  * This simplifies `LoadCAFromK8s` by making it only load the CA and not both get the secret data from K8s and load the CA.
  * This also simplifies checking whether the CA exists.
* Updates `StoreCAInK8s` to return the pointer for the secret that is created. Since this function already has the pointer for the new secret it is easier to return the pointer and then pass it to `LoadCAFromK8s` when a new CA is automatically generated.
* Updates clients of `LoadCAFromK8s` and `StoreCAInK8s` to include the new `GetCASecret` function and new dependencies.

Signed-off-by: Nicholas Lane <nicklaneovi@gmail.com>

### CODE REVIEW

File: clustermesh/certs.go
Comment (soggiest) [reply clustermesh/certs.go:187]: Sounds good to me. I had initially intended to have all of that under the `GetCASecret` function but couldn't use the logger to tell the user we were generating the certificate in CertManager. I've changed the logging message after `LoadCAFromK8s` to accommodate both of these functions (since success of both of those functions are really all a user cares about) and moved the cert generation code into `GetCASecret`.

**Review by ldelossa — APPROVED** [review]
nice. LGTM

**Comment by ldelossa** [comment]
@soggiest that merge commit should probably go away. Also would you mind adding commit descriptions to the commits in this pull? Make sure each one is signed off for the bots to be appeased :robot:

**Comment by soggiest** [comment]
> @soggiest that merge commit should probably go away. Also would you mind adding commit descriptions to the commits in this pull? Make sure each one is signed off for the bots to be appeased 🤖

Yup, I made the mistake of clicking the "Update Branch" button in the "Merging is blocked" field. I'll rebase shortly.

**Comment by soggiest** [comment]
@ldelossa Would you mind re-approving? Looks like with that rebase GH is requiring a new approval.

**Comment by ldelossa** [comment]
@soggiest not necessary, my review was not dismissed. Assigned the appropriate label and now the PR is eligible for merging.

**Comment by tklauser** [comment]
This still PR requires review from someone on the @cilium/hubble team (I see @kaworu and @glibsm currently assigned), so I've removed the `ready-to-merge` label again. The bot should add the label automatically once all necessary reviews are in.

**Comment by soggiest** [comment]
Roger that, I wasn't aware how this all works for this repo. Sorry for the noise, y'all.

File: clustermesh/certs.go
Comment (glibsm) [inline clustermesh/certs.go:189]: I think this log message might be misleading. It's not the CA it's the CA secret, right?

File: clustermesh/certs.go
Comment (glibsm) [inline clustermesh/certs.go:199]: Since the function that loads the secret doesn't take this as a param, sooner or later this log line is going to break and be out of date. It would be better to accept `defaults.CASecretName` as a param to `GetCASecret`

**Review by glibsm — CHANGES_REQUESTED** [review]

File: clustermesh/certs.go
Comment (soggiest) [reply clustermesh/certs.go]: I'm all for being more verbose. I've updated this log message

File: clustermesh/certs.go
Comment (soggiest) [reply clustermesh/certs.go:199]: Updated GetCASecret to include this parameter

**Review by glibsm — APPROVED** [review]

File: internal/certs/certs.go
Comment (soggiest) [reply internal/certs/certs.go:55]: I did this to allow `StoreCAInK8s` to return a properly namespaced `s` variable. Check out line 63.

File: hubble/ui.go
Comment (soggiest) [reply hubble/ui.go:191]: I suspect this is a Github UI bug, this value is currently in [master](https://github.com/cilium/cilium-cli/blob/master/hubble/ui.go#L191)

Comment (tklauser) [reply hubble/ui.go:191]: I also see the commit with a different SHA than on `master` when checking out the branch locally. I think this should be resolved once you rebase.

File: internal/certs/certs.go
Comment (tklauser) [reply internal/certs/certs.go:55]: I think you can still use the short variable declaration here and then reuse the variables when calling `c.StoreCAInK8s(ctx)`. It's using a "standard" assignment there (i.e. `s, err = c.StoreCAInK8s(ctx)`) which will just overwrite `s` and `err` regardless of whether they were declared explicitly or using short variable declaration on assignment.

File: hubble/hubble.go
Comment (kaworu) [inline hubble/hubble.go:251]:
```suggestion
		k.Log("❌ Unable to get or create Cilium CA Secret: %s", err)
```

File: clustermesh/certs.go
Comment (kaworu) [inline clustermesh/certs.go:198]: nitpicking
```suggestion
		k.Log("🔑 Found CA in secret %s", caSecret.Name)
```

File: hubble/hubble.go
Comment (kaworu) [inline hubble/hubble.go:261]: nitpicking
```suggestion
		k.Log("🔑 Found CA in secret %s", caSecret.Name)
```

File: install/certs.go
Comment (kaworu) [inline install/certs.go:99]:
```suggestion
		k.Log("❌ Unable to get or create Cilium CA Secret: %s", err)
```

File: install/certs.go
Comment (kaworu) [inline install/certs.go:108]: nitpicking
```suggestion
				if err := k.client.DeleteSecret(ctx, k.params.Namespace, caSecret.Name, metav1.DeleteOptions{}); err != nil {
					k.Log("Cannot delete %s Secret: %s", caSecret.Name, err)
```

File: install/certs.go
Comment (kaworu) [inline install/certs.go:113]: nitpicking
```suggestion
		k.Log("🔑 Found CA in secret %s", caSecret.Name)
```

File: internal/certs/certs.go
Comment (kaworu) [inline internal/certs/certs.go:51]: nitpicking
```suggestion
// GetOrCreateCASecret returns the k8s Secret data for the Cilium CA. If the
// Cilium CA does not already exists, this function will generate a new one
// when createCA is true.
```

**Review by kaworu — CHANGES_REQUESTED** [review]
Thank you for the PR @soggiest! I have a few nitpicking comments but overall the patch LGTM.

**Review by tklauser — APPROVED** [review]
Thanks @soggiest! LGTM once Alex' review comments are addressed.

**Review by kaworu — APPROVED** [review]
Thank you!

**Comment by kaworu** [comment]
To summarize, is it fair to say that the change from this PR is to default `--create-ca` to `true` (the rest of the patch being refactoring)? If correct, should we change the PR's title to something like `clustermesh/hubble: make --create-ca true by default` for the release note? @tklauser

**Comment by tklauser** [comment]
Good point @kaworu. I've changed the title to `clustermesh, hubble: generate Cilium CA by default` which is a mix between the former title and your suggestion 😄

**Pull request merged**
